Added query specifications

Grasmus opened the pull request.

AndriiChepeliuk committed:
Created an endpoint of obtaining a specific offer

Grasmus committed:
Added query specifications

Grasmus committed:
Added query specifications

Grasmus committed:
Added query specifications

Grasmus committed:
Added query specifications

YevheniiVovk commented on Core/Interfaces/CustomService/IRoleService.cs:
```diff
@@ -1,10 +1,11 @@
 {
     public interface IRoleService
     {
-        IdentityRole GetIdentityRoleByName(string roleName);
-        int GetRoleByName(string name);
+        Task<IdentityRole> GetIdentityRoleByNameAsync(string roleName);
+        Task<int> GetRoleByNameAsync(string name);
     }
```
Already exist https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.rolemanager-1?view=aspnetcore-6.0

YevheniiVovk replied:
There are NO NEEDS for make own service for roles

Grasmus replied:
We have two different role entities. One for Identity, second for offers

YevheniiVovk commented on Core/Interfaces/IRepository.cs:
```diff
@@ -15,5 +15,6 @@ public interface IRepository<TEntity> where TEntity : class
         Task UpdateAsync(TEntity entityToUpdate);
         Task<int> SaveChangesAsync();
         IQueryable<TEntity> Query(params Expression<Func<TEntity, object>>[] includes);
```
Why object ?

YevheniiVovk commented on Core/Interfaces/ISpecification.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public interface ISpecification<TEntity>
+    {
+        Expression<Func<TEntity, bool>> Criteria { get; }
+        IList<Expression<Func<TEntity, object>>> Includes { get; }
+        Expression<Func<TEntity, object>> OrderBy { get; }
+        Expression<Func<TEntity, object>> OrderByDescending { get; }
```
Why object? 
Why we have his interface ?

YevheniiVovk commented on Core/Interfaces/ISpecification.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Core.Interfaces
+{
+    public interface ISpecification<TEntity>
```
Let's use library

YevheniiVovk commented on Core/Services/AuthenticationService.cs:
```diff
@@ -73,12 +74,11 @@ public async Task<UserAutorizationDTO> LoginAsync(UserLoginDTO data)
-            var refreshToken = _refreshTokenRepository.Query().SingleOrDefault(t => t.Token == userLogoutDTO.RefreshToken);
+            var refreshToken = (await _refreshTokenRepository
+                .FindWithSpecificationAsync(new GetRefreshTokenByToken(userLogoutDTO.RefreshToken)))
+                .First();
```
We should NOT load the list into memory and take first element, we must load first element from db.

YevheniiVovk commented on Core/Services/AuthenticationService.cs:
```diff
@@ -118,7 +118,12 @@ private async Task<RefreshToken> CreateRefreshToken(string userId)
-            var refreshToken = GetRefreshToken(userTokensDTO.RefreshToken);
+            var refreshToken = (await _refreshTokenRepository
+                .FindWithSpecificationAsync(new GetRefreshTokenByToken(userTokensDTO.RefreshToken)))
+                .First();
```
Same

YevheniiVovk commented on Infrastructure/SpecificationEvaluator/SpecificationEvaluator.cs:
```diff
@@ -0,0 +1,34 @@
+
+            if (specification.Criteria != null)
+            {
+                query = await Task.Run(() => query.Where(specification.Criteria));
```
For a what reason we create the task ?

YevheniiVovk commented on Infrastructure/SpecificationEvaluator/SpecificationEvaluator.cs:
```diff
@@ -0,0 +1,34 @@
+
+            if (specification.OrderBy != null)
+            {
+                query = await Task.Run(() => query.OrderByDescending(specification.OrderBy));
```
For a what reason we create the task ?

YevheniiVovk commented on Infrastructure/SpecificationEvaluator/SpecificationEvaluator.cs:
```diff
@@ -0,0 +1,34 @@
+
+            if (specification.OrderByDescending != null)
+            {
+                query = await Task.Run(() => query.OrderByDescending(specification.OrderByDescending));
```
For a what reason we create the task ?

YevheniiVovk commented on Infrastructure/SpecificationEvaluator/SpecificationEvaluator.cs:
```diff
@@ -0,0 +1,34 @@
+                query = await Task.Run(() => query.OrderByDescending(specification.OrderByDescending));
+            }
+
+            query = await Task.Run(() => specification.Includes.Aggregate(query, (current, include) => current.Include(include)));
```
For a what reason we create the task ?

YevheniiVovk commented on Infrastructure/Infrastructure.csproj:
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Why this file  changed ?

Grasmus replied:
I didnt change this file. I think it`s a VS issue

YevheniiVovk commented on Core/Services/RoleService.cs:
```diff
@@ -20,16 +22,21 @@ public class RoleService : IRoleService
-            var role = _roleRepository.Query().FirstOrDefault(role => role.Name == roleName.ToUpper());
+            var role = (await _roleRepository
+                .FindWithSpecificationAsync(new GetRoleByName(roleName)))
+                .First();
```
We should not load the list into memory and then take the firs element

YevheniiVovk commented on Core/Services/RoleService.cs:
```diff
@@ -20,16 +22,21 @@ public class RoleService : IRoleService
-            var role = _identityRoleRepository.Query().FirstOrDefault(r => r.Name == roleName);
+            var role = (await _identityRoleRepository
+                .FindWithSpecificationAsync(new GetIdentityRoleByUserRoleName(roleName)))
+                .First();
```
We should not load the list into memory and then take the firsе element

YevheniiVovk commented on Core/Specifications/BaseSpecification.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Core.Specifications
+{
+    public class BaseSpecification<TEntity> : ISpecification<TEntity>
+    {
+        public BaseSpecification()
+        {
+
+        }
+
+        public BaseSpecification(Expression<Func<TEntity, bool>> criteria)
+        {
+            Criteria = criteria;
+        }
+
+        public Expression<Func<TEntity, bool>> Criteria { get; }
+        public IList<Expression<Func<TEntity, object>>> Includes { get; } = new List<Expression<Func<TEntity, object>>>();
+        public Expression<Func<TEntity, object>> OrderBy { get; private set; }
+        public Expression<Func<TEntity, object>> OrderByDescending { get; private set; }
+
+        protected void AddInclude(Expression<Func<TEntity, object>> includeExpression)
+        {
+            Includes.Add(includeExpression);
+        }
+
+        protected void AddOrderBy(Expression<Func<TEntity, object>> orderByExpression)
+        {
+            OrderBy = orderByExpression;
+        }
+
+        protected void AddOrderByDescending(Expression<Func<TEntity, object>> orderByDescendingExpression)
+        {
+            OrderByDescending = orderByDescendingExpression;
+        }
+    }
+}
```
Let's use Ardalis.Specification or some other library

AndriiChepeliuk committed:
removed ignoring rules in OfferProfiles

Eugene-pas committed:
Create endpoint to Edit user information

Eugene-pas committed:
Create endpoint to Edit user information

marynakernychna committed:
Change the relationship between Point and Trip entities

AndriiChepeliuk committed:
ID added to OfferInfoDTO

marynakernychna committed:
Change Create Point DTO & everything related with

---Main---
> Change Create Point DTO
> & everything related with it

---Others---
> Delete data return at the endpoint

AndriiChepeliuk committed:
renamed method, search Offer by Query

marynakernychna committed:
Create endpoint to create trip

marynakernychna committed:
Delete public from car interface

marynakernychna committed:
Add trip mapper

Grasmus committed:
Updated specifications with Ardalis specification

Grasmus committed:
Updated specifications with Ardalis specification

Grasmus committed:
Updated specifications with Ardalis specification

Grasmus committed:
Updated specifications with Ardalis specification

Eugene-pas committed:
problem solving

Grasmus committed:
Removed unnecessary code
Deleted unnecessary ISpecification file

Grasmus committed:
Removed unnecessary code
Deleted unnecessary ISpecification file

Grasmus committed:
Removed unnecessary code
Deleted unnecessary ISpecification file

Grasmus committed:
Removed unnecessary code
Deleted unnecessary ISpecification file

Grasmus committed:
Merge branch 'dev' into feat/dv-50/add_query_specifications

AndriiChepeliuk committed:
updated query to find offer

AndriiChepeliuk committed:
updated query to find offer

AndriiChepeliuk committed:
updated query to find offer

AndriiChepeliuk committed:
removed checking by null (entities Point, Role, GoodCategory), added async in method

Eugene-pas committed:
problem solving

Eugene-pas committed:
problem solving

Eugene-pas committed:
problem solving

AndriiChepeliuk committed:
AutoMapper was modified

Grasmus committed:
Merge branch 'dev' into feat/dv-50/add_query_specifications

marynakernychna committed:
Add trip's date validaion

marynakernychna committed:
Merge branch 'dev' into feat/dv-81/create_endpoint_to_create_trip

AndriiChepeliuk committed:
Merge pull request #30 from IT-Academy-Social-Projects-KRV/feat/dv-72/create_get_concrete_offer_endpoin

Created an endpoint of obtaining a specific offer

marynakernychna committed:
Add car check

A

marynakernychna committed:
Rename DTO back

marynakernychna committed:
Merge branch 'dev' into feat/dv-81/create_endpoint_to_create_trip

marynakernychna committed:
Supplement the ValidateTripDate method

YevheniiVovk commented on Core/Specifications/GetRoleByName.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using Ardalis.Specification;
+using Core.Entities.RoleEntity;
+
+namespace Core.Specifications
+{
+    public class GetRoleByName : Specification<Role>, ISingleResultSpecification<Role>
+    {
+        public GetRoleByName(string roleName)
+        {
+            Query.Where(r => r.Name == roleName.ToUpper());
+        }
+    }
+}
```
Let's aggregate query for each type. Like Specification/<EntityNameClass(It should be static)>.InnerClassNamedByQuery

YevheniiVovk commented on Core/Services/AuthenticationService.cs:
```diff
@@ -110,23 +107,25 @@ private async Task<RefreshToken> CreateRefreshToken(string userId)
+            var refreshToken = await _refreshTokenRepository
+                .GetBySpecAsync(new GetRefreshTokenByToken(userTokensDTO.RefreshToken));
+
+            ExceptionMethods.RefreshTokenNullCheck(refreshToken);
```
Is it a good way to check information and make an exception handling ?

Grasmus replied:
We will change it in another pr

YevheniiVovk commented on Core/Services/AuthenticationService.cs:
```diff
@@ -73,15 +73,12 @@ public async Task<UserAutorizationDTO> LoginAsync(UserLoginDTO data)
-            {
-                throw new HttpException(ErrorMessages.InvalidToken, HttpStatusCode.NotFound);
-            }
+            ExceptionMethods.RefreshTokenNullCheck(refreshToken);
```
Same

YevheniiVovk commented on Core/Services/CarService.cs:
```diff
@@ -67,12 +67,11 @@ public IQueryable<CarDTO> GetAllUserCars(string userId)
-                    c.RegistrationNumber == newCar.RegistrationNumber ||
-                    c.Vin == newCar.Vin ||
-                    c.TechnicalPassport == newCar.TechnicalPassport);
+            return await _carRepository
+                .AnyAsync(new GetCarWithMainCredentials(
+                    newCar.RegistrationNumber,
+                    newCar.Vin,
+                    newCar.TechnicalPassport));;
```
Incorrect. Please make PK(RegistrationNumber, Vin, TechnicalPassport) if all these 3 fields make row record unique

Grasmus replied:
We will make this in further pr`s

YevheniiVovk commented on Core/Services/GoodCategoryService.cs:
```diff
@@ -26,15 +25,18 @@ public class GoodCategoryService : IGoodCategoryService
+            var goodCategory = await _goodCategoryRepository
+                .GetBySpecAsync(new GetGoodCategoryByName(goodCategoryName));
+
             ExceptionMethods.GoodCategoryNullCheck(goodCategory);
```
Same as bellow. This is the bad way to make exception handling.

marynakernychna committed:
Add additional car check

marynakernychna committed:
Merge pull request #35 from IT-Academy-Social-Projects-KRV/feat/dv-81/create_endpoint_to_create_trip

Create endpoint to create trip

Eugene-pas committed:
code fix

Eugene-pas committed:
troubleshooting and formatting

Eugene-pas committed:
Merge branch 'dev' into feat/dv-62/create_endpoint_to_Edit_user_information

Eugene-pas committed:
Merge pull request #34 from IT-Academy-Social-Projects-KRV/feat/dv-62/create_endpoint_to_Edit_user_information

Feat/dv 62/create endpoint to edit user information

Grasmus committed:
Specifiaction refactoring

Grasmus committed:
Specifiaction refactoring

Grasmus committed:
Rebased branch

Grasmus committed:
Added specifications
Added migration with renamed OfferRole

Grasmus committed:
Added specifications
Added migration with renamed OfferRole

Grasmus commented on Core/Interfaces/CustomService/IOfferRoleService.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IOfferRoleService
+    {
+        //Task<IdentityRole> GetIdentityRoleByNameAsync(string roleName);
```
Will be removed in the next commit

Grasmus committed:
Added specifiactions. renamed role entities

Grasmus committed:
Resolved conflicts

Grasmus committed:
Resolved AddSeed migration conflicts

Grasmus committed:
Added seeds

vladtymo commented on Core/Services/TripService.cs:
```diff
@@ -0,0 +1,77 @@
+            await _tripRepository.AddAsync(trip);
+        }
+
+        private async Task ValidateTripDate(DateTimeOffset startDate, DateTimeOffset expirationDate, string creatorId)
```
Why don't you put this method next to the others which throw exceptions based on validation results...?

vladtymo commented on Core/Services/CarService.cs:
```diff
@@ -38,41 +38,47 @@ public class CarService : ICarService
             return _mapper.Map<CarDTO>(car);
         }
 
         public IQueryable<CarDTO> GetAllUserCars(string userId)
```
This method doesn't load the data from db into memory because of IQueryable storing data out-of-memory. Is it OK?

YevheniiVovk commented on Core/Services/TripService.cs:
```diff
@@ -0,0 +1,77 @@
+
+        public async Task CreateTripAsync(CreateTripDTO createTripDTO, string creatorId)
+        {
+            await ValidateTripDate(createTripDTO.StartDate, createTripDTO.ExpirationDate, creatorId);
+
+            var isCarValid = await _carService
+                                .CheckIsCarBelongsToUserByIds(createTripDTO.TransportationCarId, creatorId);
+
+            if (!isCarValid)
+            {
+                throw new HttpException(ErrorMessages.CarNotFound, HttpStatusCode.NotFound);
+            }
+
+            var isCarVerified = await _carService.CheckIsCarVerifiedById(createTripDTO.TransportationCarId);
+
+            if (!isCarVerified)
+            {
+                throw new HttpException(ErrorMessages.CarIsNotVerified, HttpStatusCode.BadRequest);
+            }
+
+            var trip = _mapper.Map<Trip>(createTripDTO);
+
+            trip.IsActive = false;
+            trip.IsEnded = false;
+            trip.TripCreatorId = creatorId;
+
+            await _tripRepository.AddAsync(trip);
+        }
```
Every thing is HTTPException - bad idea

marynakernychna reviewed (approved):
✨

Eugene-pas reviewed (approved):
Where any comments do not relate to the task of the Vlad. They will be resolved in the future.
